src/upgrade-root/upgrade-root.circom

```diff
@@ -0,0 +1,76 @@
+include "../../node_modules/circomlib/circuits/comparators.circom";
+include "../../node_modules/circomlib/circuits/poseidon.circom";
+include "../../node_modules/circomlib/circuits/smt/smtprocessor.circom";
+include "../../node_modules/circomlib/circuits/smt/smtverifier.circom";
+include "../../node_modules/circomlib/circuits/sha256/sha256.circom";
+
+include "./upgrade-step.circom"
+
+/**
+ * Upgrade state root account by account
+ * @param nLevels - merkle tree depth
+ * @param nAccounts - accounts to update
+ * @input initialIdx - {Uint48} - first account to update
+ * @input maxIdx - {Field} - maximum account to update
+ * @input oldStateRoot - {Field} - initial state root
+ * @input tokenID[nAccounts] - {Array(Uint32)} - tokenID of the account to update
+ * @input nonce[nAccounts] - {Array(Uint40)} - nonce of the account to update
+ * @input sign[nAccounts] - {Array(Bool)} - sign of the account to update
+ * @input balance[nAccounts] - {Array(Uint192)} - balance of the account to update
+ * @input ay[nAccounts] - {Array(Field)} - ay of the account to update
+ * @input ethAddr[nAccounts] - {Array(Uint160)} - ethAddr of the account to update
+ * @input siblings[nAccounts][nLevels + 1] - {Array[Array(Field)]} - siblings merkle proof account to update
+ * @output finalIdx - {Uint48} - last account that has been updated
+ * @output newStateRoot - {Field} - last state root after account updating
+ */
+template UpgradeRoot(nLevels, nAccounts) {
+
+    signal input initialIdx;
```

Comment:
isn't better to have just one hashed output?

Reply:
I think it is better to not have one single public input.
We use this technique (hashing all pretended public inputs-outputs) in order to save gas when calling the verifier in functionalities like `forgeBatch` and `withdrawal` since they should be called a lot of times while the roillup is living. The trade off is that we increase the number of constraints in the circuit.

In this case, `upgrade-root`, should be called just a few times to upgrade the root and it will not be called again, so gas consumption is not an issue. Besides, all the constraints saved by not including the hash of all pretended input could be used to add more accounts to upgrade to the circuit, which will decrease the total number of Ethereum calls i order to upgrade the root.